domain/miningmanager/mempool/mempool_utxo_set.go

```diff
@@ -0,0 +1,76 @@
+package mempool
+
+import (
+	"fmt"
+
+	"github.com/kaspanet/kaspad/domain/consensus/utils/consensushashing"
+
+	"github.com/kaspanet/kaspad/domain/consensus/model/externalapi"
+	"github.com/kaspanet/kaspad/domain/consensus/utils/utxo"
+	"github.com/kaspanet/kaspad/domain/miningmanager/mempool/model"
+)
+
+type mempoolUTXOSet struct {
+	mempool                       *mempool
+	poolUnspentOutputs            model.OutpointToUTXOEntry
+	transactionByPreviousOutpoint model.OutpointToTransaction
+}
+
+func newMempoolUTXOSet(mp *mempool) *mempoolUTXOSet {
+	return &mempoolUTXOSet{
+		mempool:                       mp,
+		poolUnspentOutputs:            model.OutpointToUTXOEntry{},
+		transactionByPreviousOutpoint: model.OutpointToTransaction{},
+	}
+}
+
+// this function MUST be called with the mempool mutex locked for writes
+func (mpus *mempoolUTXOSet) addTransaction(transaction *model.MempoolTransaction) {
+	outpoint := &externalapi.DomainOutpoint{TransactionID: *transaction.TransactionID()}
+
+	for i, input := range transaction.Transaction().Inputs {
+		outpoint.Index = uint32(i)
+
+		delete(mpus.poolUnspentOutputs, *outpoint)
+		mpus.transactionByPreviousOutpoint[input.PreviousOutpoint] = transaction
+	}
+
+	for i, output := range transaction.Transaction().Outputs {
+		outpoint := externalapi.DomainOutpoint{TransactionID: *transaction.TransactionID(), Index: uint32(i)}
+
+		mpus.poolUnspentOutputs[outpoint] =
+			utxo.NewUTXOEntry(output.Value, output.ScriptPublicKey, false, model.UnacceptedDAAScore)
+	}
+}
+
+// this function MUST be called with the mempool mutex locked for writes
+func (mpus *mempoolUTXOSet) removeTransaction(transaction *model.MempoolTransaction) {
+	for _, input := range transaction.Transaction().Inputs {
+		mpus.poolUnspentOutputs[input.PreviousOutpoint] = input.UTXOEntry
+		delete(mpus.transactionByPreviousOutpoint, input.PreviousOutpoint)
```

Comment:
I think we want to check whether it existed in transactionByPreviousOutpoint in the first place as well.

Reply:
Is there any good reason for this except sanity check?

If only for sanity - I'd rather not. (For the sake of readability if nothing else)